enclave-runner/src/usercalls/interface.rs

```diff
@@ -13,149 +13,283 @@ use fortanix_sgx_abi::*;
+                        } else {
+                            ret = self.0.read_alloc(fd, &mut out).await;
+                        }
+                    }
                 }
-                self.0.read_alloc(fd, &mut out)
-            })()
-            .to_sgx_result())
+                return (self, Ok(ret.to_sgx_result()));
+            }
         }
+            .boxed_local()
     }
 
-    fn write(&mut self, fd: Fd, buf: *const u8, len: usize) -> UsercallResult<(Result, usize)> {
-        unsafe {
-            Ok(from_raw_parts_nonnull(buf, len)
-                .and_then(|buf| self.0.write(fd, buf))
-                .to_sgx_result())
+    fn write(
+        self,
+        fd: Fd,
+        buf: *const u8,
+        len: usize,
+    ) -> std::pin::Pin<Box<dyn Future<Output = (Self, UsercallResult<(Result, usize)>)> + 'future>>
+    {
+        async move {
+            unsafe {
+                let ret = match from_raw_parts_nonnull(buf, len) {
+                    Ok(buf) => self.0.write(fd, buf).await,
+                    Err(e) => Err(e.into()),
+                };
+                return (self, Ok(ret.to_sgx_result()));
+            }
         }
+            .boxed_local()
     }
 
-    fn flush(&mut self, fd: Fd) -> UsercallResult<Result> {
-        Ok(self.0.flush(fd).to_sgx_result())
+    fn flush(
+        self,
+        fd: Fd,
+    ) -> std::pin::Pin<Box<dyn Future<Output = (Self, UsercallResult<Result>)> + 'future>> {
+        async move {
+            let ret = Ok(self.0.flush(fd).await.to_sgx_result());
+            return (self, ret);
+        }
+            .boxed_local()
     }
 
-    fn close(&mut self, fd: Fd) -> UsercallResult<()> {
-        Ok(self.0.close(fd))
+    fn close(
+        self,
+        fd: Fd,
+    ) -> std::pin::Pin<Box<dyn Future<Output = (Self, UsercallResult<()>)> + 'future>> {
+        async move {
+            let ret = Ok(self.0.close(fd));
+            return (self, ret);
+        }
+            .boxed_local()
     }
 
     fn bind_stream(
-        &mut self,
+        self,
         addr: *const u8,
         len: usize,
         local_addr: *mut ByteBuffer,
-    ) -> UsercallResult<(Result, Fd)> {
-        unsafe {
-            let mut local_addr = local_addr.as_mut().map(OutputBuffer::new);
-            Ok(from_raw_parts_nonnull(addr, len)
-                .and_then(|addr| self.0.bind_stream(addr, local_addr.as_mut()))
-                .to_sgx_result())
+    ) -> std::pin::Pin<Box<dyn Future<Output = (Self, UsercallResult<(Result, Fd)>)> + 'future>>
+    {
+        async move {
+            unsafe {
+                let mut local_addr = local_addr.as_mut().map(OutputBuffer::new);
+                let addr = from_raw_parts_nonnull(addr, len).unwrap();
+                let ret = Ok(self
+                    .0
+                    .bind_stream(addr, local_addr.as_mut())
+                    .await
+                    .to_sgx_result());
```

Comment:
Please fix this odd formatting throughout the PR. This should all be in one line

Reply:
I assume `rustfmt` does this automatically because of line length restrictions.